File: erlang/hamming/src/hamming.erl
-module(hamming).

-import(lists, [filter/2, zipwith/3]).

-export([distance/2]).


distance(_Strand1, _Strand2)
  when length(_Strand1) /= length(_Strand2) -> {error, badarg};
distance(_Strand1, _Strand2) -> 
    length(filter(fun (_X) -> _X end,
                  zipwith(fun (_X, _Y) -> _X /= _Y end,
                          _Strand1,
                          _Strand2))).

File: javascript/hamming/hamming.js
//
// This is only a SKELETON file for the 'Hamming' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const compute = (dna1, dna2) => {
    if (dna1.length !== dna2.length) {
        throw new Error('strands must be of equal length');
    }

    return zipWith((a, b) => a === b, [[...dna1], [...dna2]])
        .filter((x) => !x)
        .length
    
};

const zipWith = (op, rs) => rs[0].map(
  (_, idx) => (rs.map((r) => r[idx])).reduce(op)
);
